distbench_threadpool.cc

```diff
@@ -74,51 +109,151 @@ SimpleThreadpool::~SimpleThreadpool() {
-  virtual void AddWork(std::function<void()> function);
+  ElasticThreadpool(int nb_threads);
+  ~ElasticThreadpool() override;
+  void AddWork(std::function<void()> function) override;
+  std::vector<ThreadpoolStat> GetStats() override;
 
  private:
-  static void Trampoline(void* heap_object_pointer);
-  threadpool thpool_;
+  void TaskRunner();
+  void StartWorkerThread();
+
+  size_t work_count_ = 0;
+  size_t max_idle_threads_ = 0;
+  size_t idle_threads_ = 0;
+  size_t threads_launched_ = 0;
+  size_t active_threads_ = 0;
+  mutable absl::Mutex work_mutex_;
+  absl::Notification shutdown_;
+  std::queue<std::function<void()> > work_queue_ ABSL_GUARDED_BY(work_mutex_);
 };
 
-CThreadpool::CThreadpool(int nb_threads) { thpool_ = thpool_init(nb_threads); }
+ElasticThreadpool::ElasticThreadpool(int nb_threads) {
+  {
+    absl::MutexLock m(&work_mutex_);
+    max_idle_threads_ = nb_threads;
+    ++idle_threads_;
+  }
+  StartWorkerThread();
+}
 
-CThreadpool::~CThreadpool() {
-  thpool_wait(thpool_);
-  thpool_destroy(thpool_);
+void ElasticThreadpool::StartWorkerThread() {
+  {
+    absl::MutexLock m(&work_mutex_);
+    ++threads_launched_;
+    ++active_threads_;
+  }
+  auto elastic_runner = [this]() { TaskRunner(); };
+  RunRegisteredThread("ElasticThreadPool", elastic_runner).detach();
 }
 
-void CThreadpool::AddWork(std::function<void()> function) {
-  // Copy the functor object to the heap, and pass the address of the heap
-  // object to the thread pool:
-  auto* fpointer = new std::function<void()>(function);
-  thpool_add_work(thpool_, Trampoline, fpointer);
+void ElasticThreadpool::AddWork(std::function<void()> function) {
+  auto ok_to_add = [this]() {
+    return (work_queue_.size() + 1) <= max_idle_threads_;
+  };
+  auto ok_conditions = absl::Condition(&ok_to_add);
+
+  absl::MutexLock m(&work_mutex_);
+  work_mutex_.Await(ok_conditions);
+  work_queue_.push(function);
+  ++work_count_;
 }
 
-void CThreadpool::Trampoline(void* heap_object_pointer) {
-  // Execute and delete the heap allocated copy of the functor object:
-  auto f = reinterpret_cast<std::function<void()>*>(heap_object_pointer);
-  (*f)();
-  delete f;
+void ElasticThreadpool::TaskRunner() {
+  bool first_iteration = true;
+  auto work_available = [this]() {
+    return !work_queue_.empty() || shutdown_.HasBeenNotified();
+  };
+  auto working_conditions = absl::Condition(&work_available);
+  std::function<void()> task;
+  bool need_to_grow_threadpool = false;
+  absl::Duration timeout = absl::Milliseconds(10);
+  while (1) {
+    work_mutex_.Lock();
+    if (!first_iteration) {
+      idle_threads_++;
+    }
+    first_iteration = false;
+    if (work_mutex_.AwaitWithTimeout(working_conditions, timeout)) {
+      if (work_queue_.empty()) {
+        --idle_threads_;
+        --active_threads_;
+        work_mutex_.Unlock();
+        return;
+      }
+      // Work is available....
+      need_to_grow_threadpool = work_queue_.size() >= idle_threads_;
```

Comment:
So the algorithm is to grow threads as long as there are new work items in the queue?

Reply:
This is intended to replace the code above which has no limits on the number of threads created, buttries to move the cost of launching threads into the threadpool as well, to increase the maximum ops per second that the calling code can achieve.